src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionScopeInternal.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal class ClientConnectionScopeProperties
+    {
+        public IServiceConnection ServiceConnection { get; set; }
+
+        // todo: add additional properties to flow with the scope here
+    }
+
+    internal class ServiceConnectionScopeHolder
+    {
+        private static AsyncLocal<ServiceConnectionScopeHolder> s_AsyncLocalSCC = new AsyncLocal<ServiceConnectionScopeHolder>();
+        internal static ServiceConnectionScopeHolder Current
+        {
+            get => s_AsyncLocalSCC.Value;
+            set => s_AsyncLocalSCC.Value = value;
+        }
+
+        public ClientConnectionScopeProperties Properties { get; set; }
+    }
+
+    internal class ServiceConnectionScopeInternal : IDisposable
+    {
+        readonly private bool _needCleanup = false;
+
+        public ServiceConnectionScopeInternal() : this(default)
+        {
+        }
+
+        internal ServiceConnectionScopeInternal(ClientConnectionScopeProperties properties)
+        {
+            if (ServiceConnectionScopeHolder.Current == null)
+            {
+                ServiceConnectionScopeHolder.Current = new ServiceConnectionScopeHolder() { Properties = properties };
+                _needCleanup = true;
+            }
+            else
+            {
+                Debug.Assert(properties == null,
```

Comment:
shall we allow replacing?

Reply:
This is a good question. There are properties that we want to carry through the client connection lifetime and not allow user code to change (or even look up). There are also properties that we want to flow with outgoing messages (e.g. tracing context) and various parts of the code should be able to change/restore according to nested scopes.

Reply:
Added a separate class `ServiceCommunicationScope<T>` which wraps `ServiceConnectionScopeInternal` and allows nested scopes to replace the current async local values. Derived classes could use it to carry various settings to the protocol level, e.g.:
~~~
    public class TracingScope : ServiceCommunicationScope<bool>
    {
        public TracingScope(bool trace) : base(trace)
        {
        }
    }
~~~